src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -597,18 +597,48 @@ internal bool IsEntryPointCandidate
 
         /// <summary>
         /// Checks if the method has an entry point compatible signature, i.e.
-        /// - the return type is either void or int
+        /// - the return type is either void, int, <see cref="System.Threading.Tasks.Task" />, 
+        /// or <see cref="System.Threading.Tasks.Task{int}" />
         /// - has either no parameter or a single parameter of type string[]
         /// </summary>
-        internal bool HasEntryPointSignature()
+        internal bool HasEntryPointSignature(CSharpCompilation compilation)
         {
-            if (this.IsVararg)
+
+            bool IsAsyncMainReturnType(TypeSymbol type)
+            {
+                var namedType = type as NamedTypeSymbol;
+                if (namedType == null)
+                {
+                    return false;
+                }
+                // Change this to `namedType.IsNonGenericTaskType` if you want to support "task-like" objects.
```

Comment:
Nit: in general we don't break up if / else if structures with comments like this.  Instead put them into the body, or add a meta comment to the local function.